packages/url_launcher/url_launcher_android/android/src/main/java/io/flutter/plugins/urllauncher/UrlLauncher.java

```diff
@@ -15,99 +15,122 @@
 import android.util.Log;
 import androidx.annotation.NonNull;
 import androidx.annotation.Nullable;
+import androidx.annotation.VisibleForTesting;
+import io.flutter.plugins.urllauncher.Messages.UrlLauncherApi;
+import io.flutter.plugins.urllauncher.Messages.WebViewOptions;
+import java.util.Map;
+
+/** Implements the Pigeon-defined interface for calls from Dart. */
+final class UrlLauncher implements UrlLauncherApi {
+  @VisibleForTesting
+  interface IntentResolver {
+    String getHandlerComponentName(@NonNull Intent intent);
+  }
 
-/** Launches components for URLs. */
-class UrlLauncher {
   private static final String TAG = "UrlLauncher";
-  private final Context applicationContext;
 
-  @Nullable private Activity activity;
+  private final @NonNull Context applicationContext;
+
+  private final @NonNull IntentResolver intentResolver;
+
+  private @Nullable Activity activity;
 
   /**
-   * Uses the given {@code applicationContext} for launching intents.
-   *
-   * <p>It may be null initially, but should be set before calling {@link #launch}.
+   * Creates an instance that uses {@code intentResolver} to look up the handler for intents. This
+   * is to allow injecting an alternate resolver for unit testing.
    */
-  UrlLauncher(Context applicationContext, @Nullable Activity activity) {
-    this.applicationContext = applicationContext;
-    this.activity = activity;
+  @VisibleForTesting
+  UrlLauncher(@NonNull Context context, @NonNull IntentResolver intentResolver) {
+    this.applicationContext = context;
+    this.intentResolver = intentResolver;
+  }
+
+  UrlLauncher(@NonNull Context context) {
+    this(
+        context,
+        intent -> {
+          ComponentName componentName = intent.resolveActivity(context.getPackageManager());
+          return componentName == null ? null : componentName.toShortString();
```

Comment:
I was able to narrow the code that's *not* being unit tested from "all of `UrlLauncher`" to just these two lines (which come from the old `canLaunch`). I spent some time looking at this, and didn't see how to mock enough of `context` to make `intent.resolveActivity` return a dummy `ComponentName`, so I just extracted this one small piece into something tests could replace.